src/fns.rs

```diff
@@ -741,6 +741,117 @@ pub(crate) fn levenshtein_distance<T: PartialEq, I: Iterator<Item = T>>(s1: I, s
     v0[t.len()]
 }
 
+#[defun]
+pub(crate) fn string_bytes(string: &str) -> i64 {
+    string.len() as i64
+}
+
+#[defun]
+pub(crate) fn string_lessp(string1: &str, string2: &str) -> bool {
+    for (c1, c2) in string1.chars().zip(string2.chars()) {
+        match std::cmp::Ord::cmp(&c1, &c2) {
+            std::cmp::Ordering::Less => return true,
+            std::cmp::Ordering::Greater => return false,
+            _ => {}
+        }
+    }
+    string1.len() < string2.len()
+}
+
+#[defun]
+pub(crate) fn string_version_lessp(string1: &str, string2: &str) -> bool {
+    let mut iter1 = string1.chars().peekable();
+    let mut char_len1 = 0;
+    let mut iter2 = string2.chars().peekable();
+    let mut char_len2 = 0;
+    let mut num1 = None;
+    let mut num2 = None;
+    //Peeking in order to try to match a pattern without progressing the iterator
+    while let (Some(_), Some(_)) = (iter1.peek(), iter2.peek()) {
+        let (Some(mut c1), Some(mut c2)) = (iter1.next(), iter2.next()) else {
+            panic!("Somehow despite checking beforehand, the pattern is invalid");
+        };
+        if c1.is_ascii_digit() {
+            num1 = create_number(c1, &mut iter1, &mut |c| {
+                char_len1 += 1;
+                c1 = c
+            });
+        } else {
+            char_len1 += 1;
+        }
+        if c2.is_ascii_digit() {
+            num2 = create_number(c2, &mut iter2, &mut |c| {
+                char_len2 += 1;
+                c2 = c
+            });
+        } else {
+            char_len2 += 1;
+        }
+
+        match std::cmp::Ord::cmp(&c1, &c2) {
+            std::cmp::Ordering::Less => return true,
+            std::cmp::Ordering::Greater => return false,
+            _ => {}
+        }
+    }
+
+    if let (Some(n1), Some(n2)) = (num1, num2) {
+        match n1.cmp(&n2) {
+            std::cmp::Ordering::Less => return true,
+            std::cmp::Ordering::Greater => return false,
+            _ => {}
+        }
+    }
+
+    // These loops are for getting the full count if one of the iterators is empty
+    for c in iter1 {
+        if !c.is_ascii_digit() {
+            char_len1 += 1;
+        }
+    }
+    for c in iter2 {
+        if !c.is_ascii_digit() {
+            char_len2 += 1;
+        }
+    }
+
+    char_len1 < char_len2
+}
+/// Helper function to create a number from a string iterator
+/// c: the current character
+/// iter: the iterator
+/// f: a function to update the current character. This is a callback.
+/// Returns the number if it was created, otherwise None
+/// This function is used to parse version numbers in `string-version-lessp`
+/// The callback is so that we can update the length of the string - numbers
+/// and to update the next non-ascii-numeric char
+#[inline]
+fn create_number<I: Iterator<Item = char>, F: FnMut(char)>(
+    mut c: char,
+    iter: &mut I,
+    f: &mut F,
+) -> Option<usize> {
```

Comment:
I think we can rewrite this to cut down on the complexity here. Rather then taking an `Iterator` we can take [Peekable](https://doc.rust-lang.org/std/iter/struct.Peekable.html) and then use `next_if` to only consume the next character if it is an ascii digit. That mean we won't have to deal with this awkward off by 1 issue, and we won't need to take a closure just to fix up the caller's local variables. Also since we only call this when we know `c` is a digit, we don't need to return an optional and can just return a usize.